File: src/main/kotlin/com/huawei/cangjie/psi/stubs/elements/CjPackageDirectiveElementType.java
package com.huawei.cangjie.psi.stubs.elements;

import com.huawei.cangjie.descriptors.DescriptorVisibilities;
import com.huawei.cangjie.descriptors.DescriptorVisibility;
import com.huawei.cangjie.psi.CjPackageDirective;
import com.huawei.cangjie.psi.stubs.CangJiePackageDirectiveStub;
import com.huawei.cangjie.psi.stubs.impl.CangJiePackageDirectiveStubImpl;
import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import com.intellij.util.io.StringRef;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class CjPackageDirectiveElementType extends CjStubElementType<CangJiePackageDirectiveStub, CjPackageDirective> {


    public CjPackageDirectiveElementType(@NotNull @NonNls String debugName) {
        super(debugName, CjPackageDirective.class, CangJiePackageDirectiveStub.class);
    }

    @Override
    public @NotNull CangJiePackageDirectiveStub createStub(@NotNull CjPackageDirective psi, StubElement<? extends PsiElement> parentStub) {
        return new CangJiePackageDirectiveStubImpl(parentStub, psi.getModifierVisibility());

    }

    @Override
    public void serialize(@NotNull CangJiePackageDirectiveStub stub, @NotNull StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getModifierVisibility().getName());

    }

    @Override
    public @NotNull CangJiePackageDirectiveStub deserialize(@NotNull StubInputStream dataStream, StubElement parentStub) throws IOException {
        StringRef modifierVisibility = dataStream.readName();
        DescriptorVisibility visibility = null;
        if (modifierVisibility != null) {
            visibility = DescriptorVisibilities.formName(modifierVisibility.getString());
        } else {
            visibility = DescriptorVisibilities.PRIVATE;
        }
        return new CangJiePackageDirectiveStubImpl((StubElement<?>) parentStub, visibility);

    }
}

File: src/main/kotlin/cn/cangnova/cangjie/psi/stubs/elements/CjPackageDirectiveElementType.kt
/*
 * Copyright 2024 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */
package cn.cangnova.cangjie.psi.stubs.elements

import cn.cangnova.cangjie.psi.CjPackageDirective
import cn.cangnova.cangjie.psi.stubs.CangJiePackageDirectiveStub
import cn.cangnova.cangjie.psi.stubs.impl.CangJiePackageDirectiveStubImpl
import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import org.jetbrains.annotations.NonNls
import java.io.IOException

class CjPackageDirectiveElementType(debugName: @NonNls String) :
    CjStubElementType<CangJiePackageDirectiveStub , CjPackageDirective >(
        debugName,
        CjPackageDirective::class.java,
        CangJiePackageDirectiveStub::class.java
    ) {
    override fun createStub(
        psi: CjPackageDirective,
        parentStub: StubElement<out PsiElement?>
    ): CangJiePackageDirectiveStub {
        return CangJiePackageDirectiveStubImpl(parentStub)
    }

    @Throws(IOException::class)
    override fun serialize(stub: CangJiePackageDirectiveStub, dataStream: StubOutputStream) {
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>): CangJiePackageDirectiveStub {
        return CangJiePackageDirectiveStubImpl(parentStub)
    }
}
